### PR TITLE
Implement config hot reload on file change

### PR DESCRIPTION
**Pull request by richerve**
This PR implements hot reloading of config file and also make some slight adjustments on how the config is consumed too.

Previously the config file was fully handled by Viper therefore including all possible values, both from cli, env and anything else not included there that in this case is the groups orgs mapping information. Now the config file is a separate entity and Viper only handles cli and env variables.

## New Config object

Config now have the groups mapping but protected by mutexes so accessing it through a go routine, when the file is refreshed with fsnotify, it's safe to do it concurrently. In this regard there's a `Load(file)` method that simply reads the file and fills the internal groups map without watching events and there's `Watch(file)` that uses Load underneat and tracks changes in the file.

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -1,27 +1,150 @@
...
 // present in config and returns a filtered set of Groups
-func ValidUserGroups(userGroups []string, groups Groups) Groups {
-	finalGroups := make(map[string]Group)
+func ValidUserGroups(userGroups []string, groupsMap GroupsMap) GroupsMap {
+	finalGroups := make(GroupsMap)
 
 	for _, userGroup := range userGroups {
-		if _, ok := groups[userGroup]; ok {
-			finalGroups[userGroup] = groups[userGroup]
+		if v, ok := groupsMap[userGroup]; ok {
```
Comment (whwalter) [inline pkg/config/config.go:88]: This should take []string, *config.Config and perform the look ups via Load(). Currently all lookups ignore the mutex during a reload.

File: pkg/config/config.go [hunk 2]
```diff
@@ -1,27 +1,150 @@
...
+		return err
+	}
+
+	config := &config{}
+	if err = yaml.Unmarshal(data, config); err != nil {
+		return err
+	}
+
+	// Always update internal groups when loading from file
+	for gname, gconfig := range config.Groups {
```
Comment (whwalter) [inline pkg/config/config.go:75]: This doesn't handle removed groups.

File: internal/server/server.go [hunk 3]
```diff
@@ -19,7 +19,7 @@ type GrafanaResponseHeaders struct {
 type Server struct {
 	router                 *http.ServeMux
 	cookieName             string
-	groups                 config.Groups
+	groups                 config.GroupsMap
```
[the remark this answers is not in the document]
Comment (richerve) [reply internal/server/server.go:22]: correct

File: pkg/config/config.go [hunk 4]
```diff
@@ -1,27 +1,150 @@
...
-type Org struct {
-	ID   int64  `json:"id"`
-	Role string `json:"role"`
+	return result, ok
 }
 
-// UserGroupsInConfig matches the user groups (from claims) that are
+func (c *Config) GroupsMap() GroupsMap {
+	return c.groups
+}
```
[the remark this answers is not in the document]
Comment (richerve) [reply pkg/config/config.go:61]: got it, will remove this method

File: pkg/config/config.go (hunk 2 above)
Comment (richerve) [reply pkg/config/config.go:75]: true, will add that

File: pkg/config/config.go (hunk 1 above)
Comment (richerve) [reply pkg/config/config.go:88]: I changed the logic to read the groups from the Config object directly using LoadGroup

File: pkg/config/config.go [hunk 5]
```diff
@@ -1,27 +1,179 @@
...
+
+func NewFromGroupsMap(gm GroupsMap) *Config {
+	return &Config{
+		groups: gm,
+	}
+}
+
+func (c *Config) SetGroup(name string, group Group) {
+	c.mu.Lock()
+	c.groups[name] = group
```
Comment (whwalter) [inline pkg/config/config.go:53]: This setter needs a nil check on groups to ensure it isn't a nil pointer.

Comment (richerve) [reply pkg/config/config.go:53]: sorry don't understand, groups is initialized in New() and NewFromGroupsMap()

Comment (whwalter) [reply pkg/config/config.go:53]: The groups field is private so you cannot create add it when creating a struct literal but Config is public so the following would panic:

```
	conf := config.Config{}
	fmt.Println(conf.GroupNames())
	conf.SetGroup("thing", config.Group{})
```

Yes, I should use the constructors but nothing enforces that.

Comment (richerve) [reply pkg/config/config.go:53]: true, there's no enforcement, so should I create the map in SetGroup if it's nil, no?

Comment (whwalter) [reply pkg/config/config.go:53]: If you don't want to return an error that would be the most logical option.

**Comment by richerve** [comment]
closing the PR for the moment, the branch will still be available

**Pull request closed without merging**